shared/src/main/scala/mlscript/codegen/typescript/TsTypegen.scala

```diff
@@ -0,0 +1,288 @@
+      // export const arg0: int
+      // export const f: (arg0: int) => int
+      TypegenContext(typeVarMapping, recVarSet, nonRecVarSet, typegenTermScope, typegenTypeScope)
+    }
+
+    // create a context with pre-created type var to name mapping
+    def apply(mlType: Type, typeVarMapping: MutMap[TypeVar, SourceCode]): TypegenContext = {
+      val (recVarSet, nonRecVarSet) = mlType.partitionRecTypeVarSet
+      val typegenTypeScope = Scope("localTypeScope", List.empty, typeScope)
+      val typegenTermScope = Scope("localTermScope", List.empty, termScope)
+      TypegenContext(typeVarMapping, recVarSet, nonRecVarSet, typegenTermScope, typegenTypeScope)
+    }
+  }
+
+  /** Converts a term definition to its typescript declaration including any adhoc type aliases created for it
+    *
+    * @param mlType
+    *   the type to be converted to source code
+    * @param termName
+    *   name of term that's been declared. If none is given default "res" is used to indicate type of value of result
+    *   from evaluating the term
+    * @return
+    */
+  def addTypeGenTermDefinition(mlType: Type, termName: Option[String]): Unit = {
+    // `res` definitions are allowed to be shadowed
+    val defName = termName match {
+      case Some(name) => {
+        if (termScope.existsRuntimeSymbol(name)) {
+          throw new IllFormedTsTypeError(s"A declaration with name $termName already exists.")
+        } else {
+          termScope.declareRuntimeSymbol(name)
+          name
+        }
+      }
+      case None => "res"
+    }
+
+    // Create a mapping from type var to their friendly name for lookup
+    val typegenCtx = TypegenContext(mlType)
+    val tsType = toTsType(mlType)(typegenCtx, Some(true), false);
+    // only use non recursive type variables for type parameters
+    val typeParams = typegenCtx.typeVarMapping.iterator
+      .filter(tup => typegenCtx.nonRecTypeVarSet.contains(tup._1))
+      .map(_._2)
+      .toList
+
+    typegenCode += (SourceCode(s"export declare const $defName") ++
+      SourceCode.paramList(typeParams) ++ SourceCode.colon ++ tsType)
+  }
+
+  /** Converts an mlscript type to source code representation of the ts type. It also keep tracks of extra type
+    * variables and type parameters defined through the context.
+    *
+    * polarity tracks polarity of the current type. * Some(true) - positive polarity * Some(false) - negative polarity *
+    * None - no polarity (invariant)
+    *
+    * @param mlType
+    *   mlscript type to convert
+    * @param typegenCtx
+    *   type generation context for allocating new names
+    * @param pol
+    *   polarity of type
+    * @param funcArg
+    *   true if Tuple type is the lhs of a function. Used to translate tuples as multi-parameter functions
+    * @return
+    */
+  private def toTsType(mlType: Type)(implicit
+      typegenCtx: TypegenContext,
+      pol: Option[Boolean],
+      funcArg: Boolean
```

Comment:
This probably shouldn't be an implicit parameter, as it's too easy to forget to set it back to `false` in most recursive calls. You can make it an explicit param with a default value of `false`, but make sure to consider all recursive calls carefully.

In fact, it seems you forget to do just that in most places!! The fact this is not caught in a test is concerning, and shows we probably don't have enough tests.

Please add this test and make sure the output is corrected:

```scala
:ts
def f: ((1,2),(3,4)) -> 5
//│ f: ((1, 2,), (3, 4,),) -> 5
//│  = <missing implementation>
//│ // start ts
//│ export declare const f: (arg1: (arg2: 1, arg3: 2), arg4: (arg5: 3, arg6: 4)) => 5
//│ // end ts
```